scheduling-utils/src/handshake/server.rs

```diff
@@ -0,0 +1,221 @@
+        if logon.version != VERSION {
+            return Err(HandshakeError::Version {
+                server: VERSION,
+                client: logon.version,
+            });
+        }
+
+        // Put a hard limit of 64 worker threads for now.
+        if !(1..=MAX_WORKERS).contains(&logon.worker_count) {
+            return Err(HandshakeError::WorkerCount(logon.worker_count));
+        }
+
+        // Allocate memory regions with `memfd_create` and setup queues.
+        let (tpu_to_pack_file, tpu_to_pack) = Self::create_producer(logon.tpu_to_pack_size)?;
+        let (progress_tracker_file, progress_tracker) =
+            Self::create_producer(logon.progress_tracker_size)?;
+        let (pack_to_worker_files, pack_to_worker) = (0..logon.worker_count)
+            .map(|_| {
+                let file = Self::create_shmem(logon.pack_to_worker_size)?;
+                let queue = shaq::Consumer::create(&file, logon.pack_to_worker_size)?;
+
+                Ok::<_, ShaqError>((file, queue))
+            })
+            .try_fold(
+                (Vec::default(), Vec::default()),
+                |(mut files, mut queues), res| {
+                    res.map(|(file, queue)| {
+                        files.push(file);
+                        queues.push(queue);
+
+                        (files, queues)
+                    })
+                },
+            )?;
+        let (worker_to_pack_files, worker_to_pack) = (0..logon.worker_count)
+            .map(|_| Self::create_producer(logon.worker_to_pack_size))
+            .try_fold(
+                (Vec::default(), Vec::default()),
+                |(mut files, mut queues), res| {
+                    res.map(|(file, queue)| {
+                        files.push(file);
+                        queues.push(queue);
+
+                        (files, queues)
+                    })
+                },
+            )?;
+
+        // Send the file descriptors to the client.
+        let fds: Vec<_> = [&tpu_to_pack_file, &progress_tracker_file]
+            .into_iter()
+            .chain(&pack_to_worker_files)
+            .chain(&worker_to_pack_files)
+            .map(|file| file.as_raw_fd())
+            .collect();
+        let iov = [IoSlice::new(&[LOGON_SUCCESS])];
+        let cmsgs = [ControlMessage::ScmRights(&fds)];
+        let sent =
+            socket::sendmsg::<UnixAddr>(stream.as_raw_fd(), &iov, &cmsgs, MsgFlags::empty(), None)
+                .map_err(std::io::Error::from)?;
+        debug_assert_eq!(sent, 1);
+
+        Ok(Session {
+            tpu_to_pack,
+            progress_tracker,
+            pack_to_worker,
+            worker_to_pack,
+        })
+    }
+
+    fn create_producer<T>(size: usize) -> Result<(File, shaq::Producer<T>), ShaqError> {
+        let file = Self::create_shmem(size)?;
+        let queue = shaq::Producer::create(&file, size)?;
+
+        Ok((file, queue))
+    }
+
+    fn create_shmem(size: usize) -> Result<File, ShaqError> {
```

Comment:
it would be great if we can make this work with fallback for macos since many of the devs use it;

I'm pretty sure the flow can be near identical, except we use shm_open to create the shmem instead of memfd_create? wdyt?

Reply:
Yeah I think that would work